src/st/no0/57C20.c

```diff
@@ -28,7 +28,356 @@ INCLUDE_ASM("st/no0/nonmatchings/57C20", func_us_801D7D00);
+
+            if (self->ext.et_801D8150.unk88 >= 0x480) {
+                self->ext.et_801D8150.unk88 = 0x480;
+                self->step = self->step_s;
+                if (params == 0xB && (self->step == 2 || self->step == 1)) {
+                    self->animFrameIdx = 7;
+                    self->animFrameDuration = 1;
+                    self->anim = &D_us_8018228C;
+                    self->step = 2;
+                }
+            }
+        } else {
+            self->ext.et_801D8150.unk86--;
+        }
+        break;
+    case 7:
+        if (!(g_Timer & 7)) {
+            PlaySfxPositional(SFX_EXPLODE_B);
+        }
+        self->ext.et_801D8150.unk7E += 0x180;
+        if ((params == 0xB) && !(self->ext.et_801D8150.unk86 & 1)) {
+            entity = g_api.GetFreeEntity(0xE0, 0x100);
+            if (entity != NULL) {
+                DestroyEntity(entity);
+                CreateEntityFromCurrentEntity(E_EXPLOSION, entity);
+                entity->params = 1;
+                entity->zPriority = self->zPriority + 2;
+                if (self->facingLeft) {
+                    entity->posX.i.hi =
+                        self->posX.i.hi + (rand() & 0x1F) - 0x1F;
+                } else {
+                    entity->posX.i.hi = self->posX.i.hi + (rand() & 0x1F) - 0xF;
+                }
+                entity->posY.i.hi = self->posY.i.hi + (rand() & 0x1F) - 0x10;
+            }
+        }
+        if (--self->ext.et_801D8150.unk86 == 0) {
+            self->step++;
+            self->ext.et_801D8150.unk86 = (0xC - params) * 2;
+        }
+        break;
+    case 8:
+        if (--self->ext.et_801D8150.unk86 == 0) {
+            if (params == 0xB) {
+                PlaySfxPositional(SFX_STUTTER_EXPLODE_B);
+            }
+            self->step++;
+            if (params != 0) {
+                self->posY.i.hi += 8;
+                EntityExplosionSpawn(1, 0);
+                return;
+            }
+            func_us_801D7D00(2U);
+            self->animSet = 0;
+            posX = self->posX.i.hi;
+            posY = self->posY.i.hi;
+            self->primIndex = g_api.AllocPrimitives(PRIM_GT4, 1);
+            if (self->primIndex != -1) {
+                self->flags |= FLAG_HAS_PRIMS;
+                prim = &g_PrimBuf[self->primIndex];
+                prim->tpage = 0x1A;
+                prim->clut = 0x15F;
+                prim->u0 = prim->u2 = 0;
+                prim->u1 = prim->u3 = 0x3F;
+                prim->v0 = prim->v1 = 0xC0;
+                prim->v2 = prim->v3 = 0xFF;
+                prim->drawMode = DRAW_TPAGE2 | DRAW_TPAGE | DRAW_COLORS |
+                                 DRAW_UNK02 | DRAW_TRANSP;
+                prim->priority = self->zPriority - 2;
+
+                prim->r0 = prim->g0 = prim->b0 = prim->r1 = prim->g1 =
+                    prim->b1 = prim->r2 = prim->g2 = prim->b2 = prim->r3 =
+                        prim->g3 = prim->b3 = 0x80;
```

Comment:
Minor nit, this can be simplified to `PGREY(prim, 0) = PGREY(prim, 1) = PGREY(prim, 2) = PGREY(prim, 3) = 0x80;`